### DISCUSSION TITLE
Empty HTML attributes are stripped in normalisation, is this an intended behaviour?

### DISCUSSION THREAD

**Discussion by GiovanniSalmeri** [datenstrom/community#808] [Ask a question] [answered]
I am not sure if this is an intended behaviour or a tiny bug. Presently the function `getTextAttributes` (core.php:2908), used by `normaliseData` (core.php:3471), deletes empty attributes. But in this way the semantic of so-called boolean attributes could result changed. For example:

```
<input name="a" required="">
```

which means that the field *is* required, is changed in

```
<input name="a">
```

which means that the field is *not* required. Another problem is for example with attribute `alt`, for which in some cases an empty value is [officially recommended for accessibility reasons](https://www.w3.org/WAI/tutorials/images/decision-tree/) (this is different from omitting the attribute). Considering that an empty attribute anyway does no harm, perhaps core:2944 could be changed in `if (!strempty($key))`? Thank you!

**Comment by annaesvensson** [comment, marked as answer]
Yes, this is intended behaviour.

**Reply by GiovanniSalmeri** [reply]
Thank you for the answer, I will cope with it 😊

**Reply by annaesvensson** [reply]
You don't seem too happy with this answer, maybe some more context can help.

The normalisation of the Markdown parser is intentionally strict, the benefits may not be immediately apparent, but this strictness shouldn't be a problem once you are familiar with this mindset. Here are some [software design patterns](https://en.wikipedia.org/wiki/Software_design_pattern) which we use in [extensions](https://github.com/datenstrom/yellow-extensions) to generate HTML output. A boolean attribute should have a value such as `true` or `false`, a required attribute should have a meaningful value or generate output such as `checked="checked"`, an alt attribute should fetch a descriptive caption via the API or generate one based on the file name, a decorative image element should use `aria-hidden="true"` to make its intention clear to screen readers. In special cases where empty alt text is recommended to hide repeated text, you should use [aria-hidden](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-hidden) instead. As a last resort, you can write your own parser free from any normalisation, safeguards and  restrictions. You do you.

**Reply by GiovanniSalmeri** [reply]
Sorry for my reply, I meant "I will adapt my code to this normalisation" (unfortunately I think in Italian and then I write in English, the results are not always good...). Anyway, thank you for your further explication, all is very clear!

**Reply by annaesvensson** [reply]
Thank you Giovanni for adapting your code, much appreciated. 👍

**Comment by GiovanniSalmeri** [comment]
Just two sidenotes for those who should arrive to this discussion.

1. A case in which empty attributes seem necessary is the dummy option in a select element:

```
<select name="cms" required="required">
  <option value="">None</option>
  <option>Yellow</option>
  <option>Wordpress</option>
  <option>Joomla</option>
</select>
```

If `value=""` is deleted, the attribute `required` in `<select>` has no effect and `None` becomes a possible choice. Fortunately the solution is easy: `<option label="None"></option>` has exactly the same meaning and behaviour and contains no empty attribute.

2. Should someone badly need empty attributes, Javascript Is Their Friend. It is sufficient to generate in HTML for example `attribute="[VOID]"` and add a little script along these lines (not tested):

```
document.addEventListener("DOMContentLoaded", function() {
    document.querySelectorAll("*").forEach(function(element) {
        Array.from(element.attributes).forEach(function(attribute) {
            if (attribute.value=="[VOID]") attribute.value = "";
        });
    });
}, false);
```

**Reply by markseuffert** [reply]
The HTML standard says in the [introduction](https://html.spec.whatwg.org/dev/introduction.html#a-quick-introduction-to-html) that empty strings are optional: 

> Attributes are placed inside the start tag, and consist of a name  and a value, separated by an "=" character. [...] The value, along with the "=" character, can be omitted altogether if the value is the empty string.

Later, in the [select element](https://html.spec.whatwg.org/dev/form-elements.html#the-select-element) the empty string is assigned a special meaning:

> If a select element has a required attribute specified, [...] and if the value of the first option element in the select element's list of options (if any) is the empty string, [...], then that option is the select element's placeholder label option.

Here's HTML code I used for testing this special case:

``` html
<form>
  <select name="size" required>
    <option value="">-- Select Size --</option>
    <option value="small">Small</option>
    <option value="medium">Medium</option>
    <option value="large">Large</option>
  </select>
  <br />
  <button type="submit">Submit</button>
</form>
```

Should we allow empty string for certain attributes, for example `value`?

**Reply by GiovanniSalmeri** [reply]
Yet Another HTML Inconsistence 😁

I have no strong opinion in this, since `<option label="-- Select Size --"></option>` has the same effect (tested). If this is the only case where empty attributes matter, perhaps all can be left as it is now.

**Reply by markseuffert** [reply]
I would not mind to allow empty strings for certain attributes and live with HTML's design contradictions. I think web developers will copy/paste HTML examples and the software should leave enough leeway. For example [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML) and [W3schools](https://www.w3schools.com/html/) are popular sources for HTML examples. Looking at the impact this change would have in our code base. A few lines of code need to be changed and a new `$attributesAllowEmptyString = array("alt", "value")`.

**Reply by GiovanniSalmeri** [reply]
> I think web developers will copy/paste HTML examples and the software should leave enough leeway

This is very true. If a little modification in the code base can spare some time in the brief span of web developers' human life, I vote in favour 😊

**Reply by markseuffert** [reply]
Nice way to put it. The code base has been updated as discussed.

**Reply by GiovanniSalmeri** [reply]
I think I've found another candidate for `$attributesAllowEmptyString`: the attribute `download`:

absent: do not force download  
`download`: force download with the original filename  
`download=""`: force download with the original filename  
`download="download"`: force download with the filename `download` (!)  
`download="foobar"`: force download with the filename `foobar`

(In general I do not like at all *forcing* anything, every browser has its own way for downloading the target of a link and the user should be in total control. But anyway this attribute exists and it behaves, very badly, like this.)

**Reply by markseuffert** [reply]
Good discovery, the code has been updated as you suggested.
